### PR TITLE
Extract GADT constraints from wildcard type arguments

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

**Comment by dwijnand** [comment]
This regressed with #12506 (which shipped with 3.0.2).

**Comment by dwijnand** [comment]
@abgruszecki could you review this please, or delegate the review to someone else?

**Comment by abgruszecki** [comment]
@dwijnand I was going to prepare a different fix for the problem, but it looks like it's not going to happen any time soon. Let me just suggest a comment that I think should be there and we can merge this PR.

File: compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala [hunk 1]
```diff
@@ -256,12 +256,7 @@ trait PatternTypeConstrainer { self: TypeComparer =>
               val variance = param.paramVarianceSign
               if variance != 0 && !assumeInvariantRefinement then true
               else if argS.isInstanceOf[TypeBounds] || argP.isInstanceOf[TypeBounds] then
-                // Passing TypeBounds to isSubType on LHS or RHS does the
-                // incorrect thing and infers unsound constraints, while simply
-                // returning true is sound. However, I believe that it should
-                // still be possible to extract useful constraints here.
-                // TODO extract GADT information out of wildcard type arguments
-                true
+                isSubType(SkolemType(patternTp), scrutineeTp)
```
Comment (abgruszecki) [inline compiler/src/dotty/tools/dotc/core/PatternTypeConstrainer.scala:259]:
```suggestion
                // This line was added here as a quick fix for issue #13998, 
                // to extract GADT constraints from wildcard type arguments.
                // The proper fix would involve inspecting the bounds right here and performing the correct subtyping checks, the ones that are already performed by `isSubType` below,
                // for the same reasons for which we stopped using `SkolemType` here to begin with (commit 10fe5374dc2d).
                isSubType(SkolemType(patternTp), scrutineeTp)
```

**Comment by abgruszecki** [comment]
@smarter we don't pass TypeBounds themselves to `isSubType` here, we just "give up" on what I did in the deskolemization PR (referenced in the new commit) and instead we call `isSubType` on the entire types. It's not proper in the sense that using SkolemType in this way results in problems in some edge cases, but at least it fixes the regression.

**Comment by smarter** [comment]
Do you have an example of these edge cases?

**Comment by abgruszecki** [comment]
These issues were fixed by the deskolemization PR:
- https://github.com/lampepfl/dotty/issues/11103
- https://github.com/lampepfl/dotty/issues/12476

**Comment by abgruszecki** [comment]
So after a discussion with Guillaume, there's one more thing that should be said: this PR _may_ have introduced unsoundness into the compiler. I paid too much attention to us regressing after the deskolemization PR, but that PR was actually necessary to fix some soundness issues, like https://github.com/lampepfl/dotty/issues/11103. With that in mind, we should either have a very good & convincing case why this fix is certainly sound, or we should go ahead and rewrite the `isSubType` check added in this PR to compare the "elements" of TypeBounds instead of comparing the entire types.

**Comment by dwijnand** [comment]
If you can come up with an unsound example, I'm happy to redo this PR.  But without more information I wouldn't know which parts of isSubType are the unsound bits and which are the sound ones that should be inlined.

Are you free next Tuesday to pair on this in the Scala Spree?

**Pull request merged**